PWGHF: Add Xtask, add CollisionId to candidate table

rspijkers opened the pull request:
Adds the X task to O2. This pull request replaces @ginnocen PR #5409

ginnocen committed:
Add untested version oft the X to Jpsi track track

rspijkers committed:
This commit adds the Xtask to O2

rspijkers committed:
removed the filter on aod::BigTracks

rspijkers committed:
Fixed clang indentation errors (hopefully)

rspijkers committed:
previous commit still had clang issues, this should resolve it (indentation)

rspijkers committed:
temporary commit, please ignore

rspijkers committed:
Disables TPC in J/psi selector;
Adds duplicate track check in X task.

rspijkers committed:
Fixes the duplicate track check

rspijkers committed:
Clean up X task

vkucera commented on Analysis/Tasks/PWGHF/taskX.cxx:
```diff
@@ -0,0 +1,97 @@
+/// \brief X(3872) analysis task
+///
+/// \author Gian Michele Innocenti <gian.michele.innocenti@cern.ch>, CERN
+
```
Add yourself.

vkucera commented on Analysis/Tasks/PWGHF/taskX.cxx:
```diff
@@ -0,0 +1,97 @@
+  }
+};
+
+/// X analysis task
```
X(3872) analysis task

vkucera commented on Analysis/Tasks/PWGHF/taskX.cxx:
```diff
@@ -0,0 +1,97 @@
+    {{"hmassJpsi", "2-prong candidates;inv. mass (#pi K) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0., 5.}}}},
+     {"hptcand", "X candidates;candidate #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}}}};
+
+  Configurable<int> d_selectionFlagJpsi{"d_selectionFlagJpsi", 1, "Selection Flag for Jpsi"};
```
`hMassJpsi`, `hPtCand`, `selectionFlagJpsi`

vkucera commented on Analysis/Tasks/PWGHF/taskX.cxx:
```diff
@@ -0,0 +1,97 @@
+        continue;
+      }
+      registry.fill(HIST("hmassJpsi"), InvMassJpsiToEE(candidate));
+      for (auto trackPos1 = tracks.begin(); trackPos1 != tracks.end(); ++trackPos1) {
```
`for (auto& trackPos : tracks) {`

vkucera commented on Analysis/Tasks/PWGHF/taskX.cxx:
```diff
@@ -0,0 +1,97 @@
+        if (trackPos1.signed1Pt() < 0) {
+          continue;
+        }
+        for (auto trackNeg1 = tracks.begin(); trackNeg1 != tracks.end(); ++trackNeg1) {
```
`for (auto& trackNeg : tracks) {`
Maybe Partitions might be useful here in a later version.

ginnocen replied:
Ciao Yoda @vkucera, you are right. I suggest to keep it simple for the moment (there was also an issue observed with partitions). But you are right we should consider it for the future.

rspijkers committed:
Added author and fixed variable names

rspijkers committed:
Fixed clang-format

rspijkers committed:
Various fixes:
- added home institute (UU)
- changed pion loop statement (shouldn't do anything different)
- deleted print statements
- deleted JpsiTracks, no longer needed (only used for print statements)
- edited some variable names I forgot last commit

vkucera commented on Analysis/Tasks/PWGHF/taskX.cxx:
```diff
@@ -0,0 +1,107 @@
+WorkflowSpec defineDataProcessing(ConfigContext const&)
+{
+  WorkflowSpec workflow{
+    adaptAnalysisTask<AddCollisionId>("hf-task-add-collisionId"),
```
It would be better to give this task a unique name because there is already an identical one in `taskBplus.cxx`.
I would suggest `hf-task-x-add-collisionId`.

rspijkers replied:
changed it as you suggested

rspijkers committed:
hf-task-add-collisionId --> hf-task-x-add-collisionId

rspijkers committed:
Create separate workflow for AddCollisionId
Only applied to X task for now

rspijkers committed:
Minor fixes, commit for troubleshooting

rspijkers committed:
Integrates collisionId into HFCandidateCreators
instead of making it a separate workflow
- also removes aod::Colls from both X and B Tasks
(the declaration has to be removed from B task still)
- adds cfgc to taskX.cxx

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreator2Prong.cxx:
```diff
@@ -107,7 +107,8 @@ struct HFCandidateCreator2Prong {
 
       // fill candidate table rows
-      rowCandidateBase(collision.posX(), collision.posY(), collision.posZ(),
+      rowCandidateBase(rowTrackIndexProng2.index0().collisionId(),
+                       collision.posX(), collision.posY(), collision.posZ(),
```
Use the available collision row directly:
`rowTrackIndexProng2.index0().collisionId()` -> `collision.globalIndex()`

vkucera commented on Analysis/Tasks/PWGHF/taskBPlus.cxx:
```diff
@@ -65,7 +65,7 @@ struct TaskBplus {
   Partition<aod::BigTracks> positiveTracks = aod::track::signed1Pt >= 0.f;
 
-  void process(aod::Collision const&, aod::BigTracks const&, soa::Filtered<soa::Join<aod::HfCandProng2, aod::HFSelD0Candidate, aod::Colls>> const& candidates)
+  void process(aod::Collision const&, aod::BigTracks const&, soa::Filtered<soa::Join<aod::HfCandProng2, aod::HFSelD0Candidate>> const& candidates)
```
Please remove also the table and the `AddCollisionId` task.

rspijkers committed:
minor fixes:
- use the collision row for collision ID
- removed obsolete table and task from B task
- fix indentation in HFSecondaryVertex.h

vkucera commented on Analysis/Tasks/PWGHF/HFJpsiToEECandidateSelector.cxx:
```diff
@@ -79,9 +79,10 @@ struct HFJpsiToEECandidateSelector {
     if (track.charge() == 0) {
       return false;
     }
```
Please remove this check.

rspijkers replied:
I will do that, I also commented everything that has to do with the TPC, do you want me to uncomment this before committing?

vkucera commented on Analysis/Tasks/PWGHF/taskBPlus.cxx:
```diff
@@ -15,6 +15,7 @@
 /// \author Vít Kučera <vit.kucera@cern.ch>, CERN
 /// \author Nima Zardoshti <nima.zardoshti@cern.ch>, CERN
 
+#include "Framework/runDataProcessing.h"
```
This line must stay below `customize`.

rspijkers committed:
jpsi candidate selector: 
- removed charge check
- changed TPC checks

rspijkers committed:
move runDataProcessing.h below customize

rspijkers committed:
Add collision ID check for testing purposes

To be removed before merging

rspijkers committed:
remove the collision ID check, add Y cut

- Y cut instead of eta cut on the Jpsi
- also corrected x-axis label of Jpsi mass histo

vkucera commented on Analysis/Tasks/PWGHF/taskX.cxx:
```diff
@@ -0,0 +1,87 @@
+
+      int index0jpsi = candidate.index0Id();
+      int index1jpsi = candidate.index1Id();
+      for (auto& trackPos : tracks) {
+        if (trackPos.signed1Pt() < 0) {
+          continue;
+        }
+        if (trackPos.globalIndex() == index0jpsi) {
+          continue;
+        }
+        for (auto& trackNeg : tracks) {
+          if (trackNeg.signed1Pt() > 0) {
+            continue;
+          }
+          if (trackNeg.globalIndex() == index1jpsi) {
+            continue;
+          }
```
You could reduce the number of iterations from N^2 down to N(N-1)/2 (i.e. more than twice) by iterating over all tracks in the first loop and iterating over `track2 = track1 + 1` in the second loop. You just need to check once that `track1.sign() != track2.sign().`

rspijkers committed:
Pion track loops more efficient

- also changed sign and index checks

vkucera reviewed (dismissed):
@ginnocen Looks fine to me.
@rspijkers Please fix the conflicts.

ginnocen commented:
@rspijkers there is a little conflict, can you fix it? thanks!

vkucera reviewed (dismissed):
@ginnocen Good for you?

ginnocen committed:
Update HFJpsiToEECandidateSelector.cxx

jgrosseo commented on Analysis/Tasks/PWGHF/HFJpsiToEECandidateSelector.cxx:
```diff
@@ -141,9 +138,11 @@ struct HFJpsiToEECandidateSelector {
   template <typename T>
   bool selectionPIDTPC(const T& track, int nSigmaCut)
   {
+    if (nSigmaCut > 999.) {
```
Why do we need this check?

jgrosseo reviewed (approved):
Comment will be considered later. Merging now.
